nixos/modules/services/tracing/tempo.nix

```diff
@@ -1,60 +1,546 @@
+      description = mdDoc ''
+        Where to store the head blocks while they are being appended to.
+      '';
+    };
+
+    searchEnable = mkOption {
+      type = bool;
+      default = true;
+      description = mdDoc ''
+        Enable tempo search.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = attrs;
+      default = {};
+      description = mdDoc ''
+        Extra configuration to pass to Tempo service.
         See https://grafana.com/docs/tempo/latest/configuration/ for available options.
       '';
     };
 
     configFile = mkOption {
-      type = types.nullOr types.path;
+      type = nullOr path;
       default = null;
-      description = lib.mdDoc ''
-        Specify a path to a configuration file that Tempo should use.
+      description = mdDoc ''
+        As an alternative to building a working Tempo configuration using the available
+        module options, the path to a complete Tempo configuration file may be provided.
+
+        When a path to a complete Tempo configuration file is provided, all Tempo module
+        options will be ignored in favor of the Tempo configuration file, with the
+        exception of the 'service.tempo.enable' and 'services.tempo.logLevel' options.
       '';
     };
   };
 
   config = mkIf cfg.enable {
-    # for tempo-cli and friends
-    environment.systemPackages = [ pkgs.tempo ];
+    assertions = mkIf (cfg.configFile == null) [
+      {
+        assertion =
+          if cfg.metricsGeneratorEnableServiceGraphs || cfg.metricsGeneratorEnableSpanMetrics
+          then cfg.metricsGeneratorStorageRemoteWrite != null
+          else true;
+        message = ''
+          Please specify a Prometheus metrics remote write endpoint when using Tempo metrics
+          generator services with 'services.tempo.metricsGeneratorStorageRemoteWrite'.
+        '';
+      }
+      {
+        assertion =
+          if cfg.storageTraceBackend == "s3"
+          then cfg.storageS3Bucket != null
+          else true;
+        message = ''
+          Please specify an S3 storage bucket when using the s3 storage backend with
+          with 'services.tempo.storageS3Bucket'.
+        '';
+      }
+      {
+        assertion =
+          if cfg.storageTraceBackend == "s3"
+          then cfg.storageS3Endpoint != null
+          else true;
+        message = ''
+          Please specify an S3 storage endpoint when using the s3 storage backend with
+          with 'services.tempo.storageS3Endpoint'.
+        '';
+      }
+    ];
 
-    assertions = [{
-      assertion = (
-        (cfg.settings == {}) != (cfg.configFile == null)
-      );
-      message  = ''
-        Please specify a configuration for Tempo with either
-        'services.tempo.settings' or
-        'services.tempo.configFile'.
-      '';
-    }];
+    networking.firewall.allowedTCPPorts = mkIf (cfg.configFile == null) (
+      [
+        cfg.httpListenPort # default: 3200
+        cfg.grpcListenPort # default: 9096
+      ]
```

Comment:
firewall should not be auto opened with no wait to turn off

Reply:
Done with openFirewall option